### ISSUE TITLE
CareKit with SwiftUI

### ISSUE THREAD

**Issue by drdavec** [carekit-apple/CareKit#316]
Has anyone been successful embedding CareKit views in a SwiftUI designed app?  I'm experimenting with a simple test app and get errors with unsatisfiable layout constraints.  I am wrapping a CareKit view as follows for use as a SwiftUI view.  Then including in a SwiftUI List.  The flow is generally working, but view is displayed empty and layout errors reported in the console.

// In SwiftUI content view:
List() {
	SimpleTaskViewController(taskIdentifier: "kegels")
}

// wraps a UIViewController for use as SwiftUI view
struct SimpleTaskViewController: UIViewControllerRepresentable {
	let careKitStore = (UIApplication.shared.delegate as! AppDelegate).manager
	
	var taskIdentifier: String
	
	public func makeUIViewController(context: Context) -> OCKSimpleTaskViewController<OCKStore> {
		let card = OCKSimpleTaskViewController(
			storeManager: careKitStore,
			taskIdentifier: self.taskIdentifier,
			eventQuery: OCKEventQuery(for: Date()))
		return card
	}
	
	public func updateUIViewController(_ uiViewController: OCKSimpleTaskViewController<OCKStore>, context: Context) {
	}
}

**Comment by lucaspozzi** [comment]
I'm about to start implementing CareKit in a SwiftUI app and I see few examples on it.
How did it go for you @drdavec ?

**Comment by drdavec** [comment]
I've been unsuccessful using CareKit views in SwiftUI.  I have also seen
the new documentation for using CareKit controllers and will evaluate that
capability soon.  Is the current recommendation to redevelop CareKit views
as SwiftUI views?  While the controllers are very helpful, I am also
interested in incorporating the Chart and Contact views into my SwiftUI app.

Dave

**Comment by gavirawson-apple** [comment]
There are two ways you can go about accomplishing this for your SwiftUI app:

1. Re-create the view and use just the controller, as noted in the README.

2. Create a SwiftUI View representation for the CareKit view controller.

```swift 
// Define the representation.
struct SimpleContactView: UIViewControllerRepresentable {

    let storeManager: OCKSynchronizedStoreManager

    func makeUIViewController(context: Context) -> OCKSimpleContactViewController {
        OCKSimpleContactViewController(contactID: "lexi-torres", storeManager: storeManager)
    }

    func updateUIViewController(_ viewController: OCKSimpleContactViewController, context: Context) {
    }
}

// Use the representation in your view code.
struct ContentView: View {

    let storeManager: OCKSynchronizedStoreManager

    var body: some View {
        SimpleContactView(storeManager: storeManager)
            .frame(width: 300, height: 80)
    }
}

```

**Comment by drdavec** [comment]
Thanks for your response.  In my original post for this issue, I included a
snippet showing how I was using SwiftUI Representable wrapper, but had
errors and layout issues.  I now see from your example that a .frame()
modifier must be included for each use of the wrapped view.  Thanks, this
now works for me.  As below, I was able to use GeometryReader to make the
width adaptable, but still must guess at a fixed height which is not always
correct, depending on actual content.

Thanks again, I will work with this approach for now.

NavigationView {

GeometryReader { geometry *in*

Form {

Section(header: Text("Contacts")) {

SimpleContactView(contactID: "matthew")

.frame(width: geometry.size.width - 30, height: 80)

DetailedContactView(contactID: "jane")

.frame(width: geometry.size.width - 30, height: 320)

}

}

}

.navigationBarTitle("CareKit")

}
